torii/hdl/time.py

```diff
@@ -0,0 +1,1147 @@
+
+	Parameters
+	----------
+	value : float
+		The value to truncate to ``dig`` if possible.
+
+	dig : int
+		The number of trailing digits to leave on ``value`` if any exist.
+
+	Returns
+	-------
+	int | float
+		The result of ``value`` rounded to ``dig`` if any trailing digits are found, and if none
+		are present then ``value`` is returned as an ``int``
+	'''
+
+	# TODO(aki): Should we `round()` /then/ do the `.is_integer()` check?
+	if value.is_integer():
+		return int(value)
+	return round(value, dig)
+
+def _auto_scale(value: float, unit_scale: tuple[str, ...], dig: int) -> str:
+	'''
+	Automatically scale the input value to the closest whole unit and return it as a formatted string.
+
+	Parameters
+	---------
+	value : float
+		The value to scale.
+
+	unit_scale : tuple[str, ...]
+		The scale units in order of smallest to largest
+
+	dig : int
+		The number of trailing digits to include if any
+
+	Returns
+	-------
+	str
+		The formatted string of the scaled value.
+	'''
+
+	adjusted_value = (value / ATTO)
+
+	scale = int(floor(log(adjusted_value, 1000)))
+	power = pow(1000, scale)
+	fixed = round(adjusted_value / power, dig)
+
+	return f'{_truncate(fixed, dig)}{unit_scale[scale]}'
+
+def _parse_format_spec(format_spec: str, unit_scale: tuple[str, ...]) -> tuple[str, int, bool]:
+	'''
+	Parse the input format specifier and return the supplied options or their defaults.
+
+	Parameters
+	----------
+	format_spec : str
+		The format string to parse
+
+	unit_scale : tuple[str, ...]
+		The scale units in order of smallest to largest
+
+	Returns
+	-------
+	tuple[str, int, bool]
+		Returns the identified scale unit, the number of trailing digits, and if the unit suffix should be stripped
+		or not.
+	'''
+
+	# Split the format string on `.` to see if we are doing any trailing digit formatting
+	opts = format_spec.split('.', maxsplit = 1)
+
+	# If we have exactly 2, then assume so and set the values appropriately
+	if len(opts) == 2:
+		(unit, dig) = opts
+		dig = int(dig, 10)
+	else:
+		# If not, just pull the units out and apply the default
+		unit = opts[0]
+		dig = 4
+
+	# If the `unit` format specifier starts with `#` then we want to strip the unit suffix
+	strip_suffix = False
+	if unit.startswith('#'):
+		unit = unit[1:]
+		strip_suffix = True
+
+	if unit not in ('', *unit_scale):
+		raise ValueError(
+			f'Unknown format specifier {unit}, expected an empty string or one of {" ".join(unit_scale)}'
+		)
+
+	return (unit, dig, strip_suffix)
+
+class Frequency:
+	'''
+	Represents a frequency in whole or fractional Hertz.
+
+	Parameters
+	----------
+	value: float
+		The frequency in Hertz.
+
+	Attributes
+	----------
+	period : torii.hdl.time.Period
+		Converts this frequency into a Torii Period.
+
+	gigahertz : float
+		Returns this frequency in Gigahertz as a float literal.
+
+	megahertz : float
+		Returns this frequency in Megahertz as a float literal.
+
+	kilohertz : float
+		Returns this frequency in Kilohertz as a float literal.
```

Comment:
Believe that the fully spelled out suffixed units here should start with a lower case letter to satisfy SI naming scheme requirements.